### PR TITLE
Update CUDA custom op unit tests to account for recent ORT change

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**: https://github.com/microsoft/onnxruntime/pull/3723 made a change to move the hand-crafted ORT CUDA kernels from the per-thread default stream to a compute stream that ORT created by default or a user provided compute stream. This change will require exposing an interface in custom ops to obtain ORT's compute stream or at the very least update the CUDA custom op unit tests so as to show that a user needs to use the same compute stream for both the ORT kernels (passed in via session options) and their custom kernels for synchronization. This change updates the unit tests.

**Motivation and Context**
The unit tests will work as is as they are just single node toy models but if users were to adapt existing custom op implementations for their models that will require interleaving with ORT's existing CUDA kernels, they will lack synchronization and result in garbage output. Hence, this change for illustrative purposes.

(Takeaway from https://github.com/microsoft/onnxruntime/issues/6952)

### CODE REVIEW

File: onnxruntime/test/shared_lib/utils.h [hunk 1]
```diff
@@ -156,3 +166,58 @@ struct MyCustomOpWithAttributes : Ort::CustomOpBase<MyCustomOpWithAttributes, My
  private:
   const char* provider_;
 };
+
+//Slice array of floats or doubles between [from, to) and save to output
+struct SliceCustomOpKernel {
```
Comment (hariharans29) [inline onnxruntime/test/shared_lib/utils.h:171]: Moved this for test_inference.cc

File: onnxruntime/test/shared_lib/utils.cc [hunk 2]
```diff
@@ -27,9 +31,10 @@ void MyCustomKernel::Compute(OrtKernelContext* context) {
 
   // Do computation
 #ifdef USE_CUDA
-  cuda_add(size, out, X, Y);
-  cudaStreamSynchronize(nullptr);
```
Comment (hariharans29) [inline onnxruntime/test/shared_lib/utils.cc:31]: No need this costly sync anymore - we are launching the custom op in the same stream as ORT's CUDA kernels - previously without this sync, we would have gotten result mismatches as things were not synchronized correctly

File: onnxruntime/test/shared_lib/utils.h (hunk 1 above)
Comment (skottmckay) [reply onnxruntime/test/shared_lib/utils.h:171]: May be better to rename files to custom_op_utils.h/.cc given their current content

---
In reply to: [591514397](https://github.com/microsoft/onnxruntime/pull/6971#discussion_r591514397) [](ancestors = 591514397)

File: onnxruntime/test/shared_lib/utils.cc (hunk 2 above)
Comment (skottmckay) [reply onnxruntime/test/shared_lib/utils.cc:31]: Can we leave the sync here but commented out saying this should never be necessary if things are setup correctly?

---
In reply to: [592109259](https://github.com/microsoft/onnxruntime/pull/6971#discussion_r592109259) [](ancestors = 592109259)

File: onnxruntime/test/shared_lib/test_inference.cc [hunk 3]
```diff
@@ -87,14 +87,24 @@ static void TestInference(Ort::Env& env, const std::basic_string<ORTCHAR_T>& mod
...
+    OrtCUDAProviderOptions cuda_options{
+        0,
+        OrtCudnnConvAlgoSearch::EXHAUSTIVE,
+        std::numeric_limits<size_t>::max(),
+        0,
+        true,
+        cuda_compute_stream != nullptr ? 1 : 0,
+        cuda_compute_stream != nullptr ? cuda_compute_stream : nullptr};
+    session_options.AppendExecutionProvider_CUDA(cuda_options);
 #else
```
Comment (skottmckay) [inline onnxruntime/test/shared_lib/test_inference.cc:106]: nit: May be useful to have a helper that creates the default options with optional compute stream.

**Comment by skottmckay** [comment]
Assuming doco on [https://www.onnxruntime.ai/docs/how-to/add-custom-op.html](https://www.onnxruntime.ai/docs/how-to/add-custom-op.html) will be updated separately to describe what's necessary to create a working CUDA based custom op and why.

**Review by skottmckay — APPROVED** [review]
:shipit:

File: onnxruntime/test/shared_lib/utils.h (hunk 1 above)
Comment (hariharans29) [reply onnxruntime/test/shared_lib/utils.h:171]: done

File: onnxruntime/test/shared_lib/utils.cc (hunk 2 above)
Comment (hariharans29) [reply onnxruntime/test/shared_lib/utils.cc:31]: done

File: onnxruntime/test/shared_lib/test_inference.cc (hunk 3 above)
Comment (hariharans29) [reply onnxruntime/test/shared_lib/test_inference.cc:100]: just created a helper method for now

**Comment by hariharans29** [comment]
> Assuming doco on https://www.onnxruntime.ai/docs/how-to/add-custom-op.html will be updated separately to describe what's necessary to create a working CUDA based custom op and why.

Yes, will come in a separate change

File: cmake/onnxruntime_unittests.cmake [hunk 4]
```diff
@@ -319,7 +319,9 @@ set (onnxruntime_shared_lib_test_SRC
           ${ONNXRUNTIME_SHARED_LIB_TEST_SRC_DIR}/test_model_loading.cc
           ${ONNXRUNTIME_SHARED_LIB_TEST_SRC_DIR}/test_ort_format_models.cc
           ${ONNXRUNTIME_SHARED_LIB_TEST_SRC_DIR}/utils.h
-          ${ONNXRUNTIME_SHARED_LIB_TEST_SRC_DIR}/utils.cc)
+          ${ONNXRUNTIME_SHARED_LIB_TEST_SRC_DIR}/utils.cc		  
+          ${ONNXRUNTIME_SHARED_LIB_TEST_SRC_DIR}/custom_op_utils.h
```
Comment (skottmckay) [inline cmake/onnxruntime_unittests.cmake:323]: nit: 2 tabs

**Review by skottmckay — APPROVED** [review]
:shipit:

File: cmake/onnxruntime_unittests.cmake (hunk 4 above)
Comment (hariharans29) [reply cmake/onnxruntime_unittests.cmake:323]: Will combine with another PR

**Pull request merged**